pandas/io/excel.py

```diff
@@ -679,6 +679,11 @@ def _parse_cell(cell_contents, cell_typ):
                                     **kwds)
 
                 output[asheetname] = parser.read(nrows=nrows)
+                dtypes = output[asheetname].dtypes
+                output[asheetname].replace('nan', np.nan, inplace=True)
+                output[asheetname] = output[asheetname].astype(dtypes,
+                                                               copy=False)
```

Comment:
I worry about this patch being a performance hit against `read_excel`.  The Python parser processes each of the Excel elements before placing it into a `DataFrame`.  I would look there for the fix.

Reply:
The result from `read_csv` is not the same as the one from `read_excel` with `dtype=str`.  In the former case, empties are read in as `np.nan`, whereas in the latter they are read in as the string 'nan'.

Reply:
True, though that doesn't change my opinion.   The problematic part still likely stems in the engine parsing, which would also effect `read_csv` (the more popular of the two IMO).  Thus, if we can kill two birds with one stone, that would be even better.